iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -4006,32 +4025,57 @@ STRING_HANDLE IoTHubTransport_MQTT_Common_GetHostname(TRANSPORT_LL_HANDLE handle
     return result;
 }
 
-IOTHUB_CLIENT_RESULT IoTHubTransport_MQTT_Common_SendMessageDisposition(MESSAGE_CALLBACK_INFO* message_data, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
+IOTHUB_CLIENT_RESULT IoTHubTransport_MQTT_Common_SendMessageDisposition(IOTHUB_DEVICE_HANDLE handle, IOTHUB_MESSAGE_HANDLE messageHandle, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
 {
-    (void)disposition;
-
     IOTHUB_CLIENT_RESULT result;
-    if (message_data)
+    if (handle == NULL || messageHandle == NULL)
+    {
+        /*Codes_SRS_IOTHUB_MQTT_TRANSPORT_10_001: [If messageData is NULL, IoTHubTransport_MQTT_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR. ]*/
+        LogError("Invalid argument (handle=%p, messageHandle=%p", handle, messageHandle);
+        result = IOTHUB_CLIENT_INVALID_ARG;
+    }
+    else if (disposition == IOTHUBMESSAGE_ASYNC_ACK)
+    {
+        LogError("IOTHUBMESSAGE_ASYNC_ACK is not a valid disposition value for this function");
+        result = IOTHUB_CLIENT_INVALID_ARG;
+    }
+    else
     {
-        if (message_data->messageHandle)
+        if (disposition == IOTHUBMESSAGE_ACCEPTED)
```

Comment:
Nit - I think this is moot because of my comments removing IOTHUBMESSAGE_ASYNC_ACK check in this layer above.  But I recommend the pattern of

if (...)
else if (...)
else if(...)

As much as possible, or else the nesting of else's to my mind is harder to follow.  So that'd suggest having the `else if (disposition == IOUTBMESSAGE_ACCEPTED)` as part of "root else".

Reply:
I agree 100%, except when there is a common code at the end of the block that would be common to the last else and a previous else-if. In that case this constructions keeps it cleaner. But we can chat about it later.
And indeed, the IOTHUBMESSAGE_ASYNC_ACK has been removed.